core/src/main/java/org/apache/iceberg/LocationProviders.java

```diff
@@ -167,5 +185,42 @@ private static String pathContext(String tableLocation) {
 
       return resolvedContext;
     }
+
+    @Override
+    public String dataLocation() {
+      return this.storageLocation;
+    }
+  }
+
+  static class DefaultMetadataLocationProvider implements MetadataLocationProvider {
+    private static final String METADATA_FOLDER_NAME = "metadata";
+    private String metadataLocation;
+
+    DefaultMetadataLocationProvider(String tableLocation, Map<String, String> properties) {
+      this.metadataLocation =
+          LocationUtil.stripTrailingSlash(
+              metadataLocation(
+                  Preconditions.checkNotNull(tableLocation, "table location cannot be null"),
```

Comment:
generally it is better to use `Preconditions.checkArgument(... != null, ...)` as it is better in indicating that something wrong was passed (rather than surfacing a NPE). See also https://github.com/apache/iceberg/issues/6625 for some background